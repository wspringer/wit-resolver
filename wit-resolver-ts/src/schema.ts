import { z } from "zod";

// Basic type aliases
export const TypeIdSchema = z.number();
export const InterfaceIdSchema = z.number();
export const WorldIdSchema = z.number();
export const PackageIdSchema = z.number();

// Basic types
export const TypeSchema = z.union([
  z.literal("bool"),
  z.literal("u8"),
  z.literal("u16"),
  z.literal("u32"),
  z.literal("u64"),
  z.literal("s8"),
  z.literal("s16"),
  z.literal("s32"),
  z.literal("s64"),
  z.literal("f32"),
  z.literal("f64"),
  z.literal("char"),
  z.literal("string"),
  z.literal("error-context"),
  TypeIdSchema,
]);

export const HandleSchema = z.union([
  z.object({ own: TypeIdSchema }),
  z.object({ borrow: TypeIdSchema }),
]);

export const FieldSchema = z.object({
  name: z.string(),
  type: TypeSchema,
  docs: z.string().optional(),
});

export const FlagSchema = z.object({
  name: z.string(),
  docs: z.string().optional(),
});

export const CaseSchema = z.object({
  name: z.string(),
  type: TypeSchema.optional(),
  docs: z.string().optional(),
});

export const EnumCaseSchema = z.object({
  name: z.string(),
  docs: z.string().optional(),
});

export const TypeDefKindSchema = z.union([
  z.object({ record: z.object({ fields: z.array(FieldSchema) }) }),
  z.literal("resource"),
  z.object({ handle: HandleSchema }),
  z.object({ flags: z.object({ flags: z.array(FlagSchema) }) }),
  z.object({ tuple: z.object({ types: z.array(TypeSchema) }) }),
  z.object({ variant: z.object({ cases: z.array(CaseSchema) }) }),
  z.object({ enum: z.object({ cases: z.array(EnumCaseSchema) }) }),
  z.object({ option: TypeSchema }),
  z.object({
    result: z.object({ ok: TypeSchema.optional(), err: TypeSchema.optional() }),
  }),
  z.object({ list: TypeSchema }),
  z.object({ future: TypeSchema.optional() }),
  z.object({ stream: TypeSchema.optional() }),
  z.object({ type: TypeSchema }),
]);

export const TypeOwnerSchema = z.union([
  z.object({ world: WorldIdSchema }),
  z.object({ interface: InterfaceIdSchema }),
  z.null(),
]);

export const TypeDefSchema = z.object({
  name: z.string().nullable(),
  kind: TypeDefKindSchema,
  owner: TypeOwnerSchema,
  docs: z.string().optional(),
  stability: z.string().optional(),
});

export const FunctionKindSchema = z.union([
  z.literal("freestanding"),
  z.literal("async-freestanding"),
  z.object({ method: TypeIdSchema }),
  z.object({ "async-method": TypeIdSchema }),
  z.object({ static: TypeIdSchema }),
  z.object({ "async-static": TypeIdSchema }),
  z.object({ constructor: TypeIdSchema }),
]);

export const FunctionSchema = z.object({
  name: z.string(),
  kind: FunctionKindSchema,
  params: z.array(
    z.object({
      name: z.string(),
      type: TypeSchema,
    })
  ),
  result: TypeSchema.optional(),
  docs: z.string().optional(),
  stability: z.string().optional(),
});

export const InterfaceSchema = z.object({
  name: z.string().optional(),
  types: z.record(z.string(), TypeIdSchema),
  functions: z.record(z.string(), FunctionSchema),
  docs: z.string().optional(),
  stability: z.string().optional(),
  package: PackageIdSchema.optional(),
});

export const WorldItemSchema = z.union([
  z.object({
    interface: z.object({
      id: InterfaceIdSchema,
      stability: z.string().optional(),
    }),
  }),
  z.object({ function: FunctionSchema }),
  z.object({ type: TypeIdSchema }),
]);

export const WorldSchema = z.object({
  name: z.string(),
  imports: z.record(z.string(), WorldItemSchema),
  exports: z.record(z.string(), WorldItemSchema),
  package: PackageIdSchema.optional(),
  docs: z.string().optional(),
  stability: z.string().optional(),
});

export const PackageSchema = z.object({
  name: z.string(),
  docs: z.string().optional(),
  interfaces: z.record(z.string(), InterfaceIdSchema),
  worlds: z.record(z.string(), WorldIdSchema),
});

export const ResolvedSchema = z.object({
  worlds: z.array(WorldSchema),
  interfaces: z.array(InterfaceSchema),
  types: z.array(TypeDefSchema),
  packages: z.array(PackageSchema),
});
