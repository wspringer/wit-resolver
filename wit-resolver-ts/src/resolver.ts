import { ResolvedSchema } from "./schema.js";
import { Resolved } from "./types.js";
import * as component from "./wit_resolver_wasm.js";

const WasmResolver = component.types.Resolver;

export interface Resolver {
  register(name: string, source: string): void;
  raw: () => object;
  resolve(): Resolved;
}

export function Resolver(): Resolver {
  const resolver = new WasmResolver();
  const raw = () => {
    const result = resolver.resolve();
    const parsed = JSON.parse(result);
    return parsed;
  };
  return {
    register: (name, source) => {
      resolver.register(name, source);
    },
    raw,
    resolve: () => {
      const parsed = raw();
      return ResolvedSchema.parse(parsed);
    },
  };
}
