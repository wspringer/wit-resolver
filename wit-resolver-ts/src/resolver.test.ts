import { Resolver } from "./resolver.js";
import { describe, it, expect } from "vitest";
import { ResolvedSchema } from "./schema.js";

describe("Resolver", () => {
  const resolve = (
    wit: string
  ): [object, ReturnType<typeof ResolvedSchema.safeParse>] => {
    const resolver = Resolver.create();
    resolver.register("test.wit", wit);
    const raw = resolver.raw();
    return [raw, ResolvedSchema.safeParse(raw)];
  };

  it("should handle an empty interface", () => {
    const [, resolved] = resolve(`
      package local:test;
      interface foo { }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved).toHaveProperty("data");
    expect(resolved.data).toHaveProperty("worlds");
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a function returning unit", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        bar: func();
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a function returning a string", () => {
    const [, resolved] = resolve(`
      package local:test;
      interface foo {
        bar: func() -> string;
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a function returning a result", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        bar: func() -> result<string, string>;
      }
    `);
    console.info(raw);
    console.info(resolved.error);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a function accepting a result", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        bar: func(a: result<string, string>);
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a resource", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        resource bar {
          constructor(a: string);
        }
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a resource with a method", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        resource bar {
          baz: func();
        }
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should handle an interface with a resource with a static method", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        resource bar {
          baz: static func();
        }
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should support all of the type definitions (enums, variants, etc.)", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo {
        enum bar {
          baz,
          qux
        }
        variant baz {
          qux,
          quux(string)
        }
        type foo = option<string>;
        type foos = list<foo>;
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should allow you define a world", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      interface foo-items {
        bar: func() -> list<string>;
      }
      world foo {
        import foo-items;
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });

  it("should allow you to use nested packages", () => {
    const [raw, resolved] = resolve(`
      package local:test;
      package foo:bar {
        interface foo-items {
          bar: func() -> list<string>;
        }
      }
      world foo {
        import foo:bar/foo-items;
      }
    `);
    expect(resolved.success).toBe(true);
    expect(resolved.data).toMatchSnapshot();
  });
});
