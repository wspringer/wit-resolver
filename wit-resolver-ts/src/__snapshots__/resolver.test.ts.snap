// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Resolver > should allow you define a world 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "bar": {
          "kind": "freestanding",
          "name": "bar",
          "params": [],
          "result": 0,
        },
      },
      "name": "foo-items",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo-items": 0,
      },
      "name": "local:test",
      "worlds": {
        "foo": 0,
      },
    },
  ],
  "types": [
    {
      "kind": {
        "list": "string",
      },
      "name": null,
      "owner": null,
    },
  ],
  "worlds": [
    {
      "exports": {},
      "imports": {
        "interface-0": {
          "interface": {
            "id": 0,
          },
        },
      },
      "name": "foo",
      "package": 0,
    },
  ],
}
`;

exports[`Resolver > should allow you to use nested packages 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "bar": {
          "kind": "freestanding",
          "name": "bar",
          "params": [],
          "result": 0,
        },
      },
      "name": "foo-items",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo-items": 0,
      },
      "name": "foo:bar",
      "worlds": {},
    },
    {
      "interfaces": {},
      "name": "local:test",
      "worlds": {
        "foo": 0,
      },
    },
  ],
  "types": [
    {
      "kind": {
        "list": "string",
      },
      "name": null,
      "owner": null,
    },
  ],
  "worlds": [
    {
      "exports": {},
      "imports": {
        "interface-0": {
          "interface": {
            "id": 0,
          },
        },
      },
      "name": "foo",
      "package": 1,
    },
  ],
}
`;

exports[`Resolver > should handle an empty interface 1`] = `
{
  "interfaces": [
    {
      "functions": {},
      "name": "foo",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a function accepting a result 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "bar": {
          "kind": "freestanding",
          "name": "bar",
          "params": [
            {
              "name": "a",
              "type": 0,
            },
          ],
        },
      },
      "name": "foo",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [
    {
      "kind": {
        "result": {
          "err": "string",
          "ok": "string",
        },
      },
      "name": null,
      "owner": null,
    },
  ],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a function returning a result 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "bar": {
          "kind": "freestanding",
          "name": "bar",
          "params": [],
          "result": 0,
        },
      },
      "name": "foo",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [
    {
      "kind": {
        "result": {
          "err": "string",
          "ok": "string",
        },
      },
      "name": null,
      "owner": null,
    },
  ],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a function returning a string 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "bar": {
          "kind": "freestanding",
          "name": "bar",
          "params": [],
          "result": "string",
        },
      },
      "name": "foo",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a function returning unit 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "bar": {
          "kind": "freestanding",
          "name": "bar",
          "params": [],
        },
      },
      "name": "foo",
      "package": 0,
      "types": {},
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a resource 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "[constructor]bar": {
          "kind": {
            "constructor": 0,
          },
          "name": "[constructor]bar",
          "params": [
            {
              "name": "a",
              "type": "string",
            },
          ],
          "result": 1,
        },
      },
      "name": "foo",
      "package": 0,
      "types": {
        "bar": 0,
      },
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [
    {
      "kind": "resource",
      "name": "bar",
      "owner": {
        "interface": 0,
      },
    },
    {
      "kind": {
        "handle": {
          "own": 0,
        },
      },
      "name": null,
      "owner": null,
    },
  ],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a resource with a method 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "[method]bar.baz": {
          "kind": {
            "method": 0,
          },
          "name": "[method]bar.baz",
          "params": [
            {
              "name": "self",
              "type": 1,
            },
          ],
        },
      },
      "name": "foo",
      "package": 0,
      "types": {
        "bar": 0,
      },
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [
    {
      "kind": "resource",
      "name": "bar",
      "owner": {
        "interface": 0,
      },
    },
    {
      "kind": {
        "handle": {
          "borrow": 0,
        },
      },
      "name": null,
      "owner": null,
    },
  ],
  "worlds": [],
}
`;

exports[`Resolver > should handle an interface with a resource with a static method 1`] = `
{
  "interfaces": [
    {
      "functions": {
        "[static]bar.baz": {
          "kind": {
            "static": 0,
          },
          "name": "[static]bar.baz",
          "params": [],
        },
      },
      "name": "foo",
      "package": 0,
      "types": {
        "bar": 0,
      },
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [
    {
      "kind": "resource",
      "name": "bar",
      "owner": {
        "interface": 0,
      },
    },
  ],
  "worlds": [],
}
`;

exports[`Resolver > should support all of the type definitions (enums, variants, etc.) 1`] = `
{
  "interfaces": [
    {
      "functions": {},
      "name": "foo",
      "package": 0,
      "types": {
        "bar": 0,
        "baz": 1,
        "foo": 2,
        "foos": 3,
      },
    },
  ],
  "packages": [
    {
      "interfaces": {
        "foo": 0,
      },
      "name": "local:test",
      "worlds": {},
    },
  ],
  "types": [
    {
      "kind": {
        "enum": {
          "cases": [
            {
              "name": "baz",
            },
            {
              "name": "qux",
            },
          ],
        },
      },
      "name": "bar",
      "owner": {
        "interface": 0,
      },
    },
    {
      "kind": {},
      "name": "baz",
      "owner": {
        "interface": 0,
      },
    },
    {
      "kind": {
        "option": "string",
      },
      "name": "foo",
      "owner": {
        "interface": 0,
      },
    },
    {
      "kind": {
        "list": 2,
      },
      "name": "foos",
      "owner": {
        "interface": 0,
      },
    },
  ],
  "worlds": [],
}
`;
